### ISSUE TITLE
[css-nav-1] Starting point isn't changed when searching outside of a container

### ISSUE THREAD

**Issue by frivoal** [w3c/csswg-drafts#3392]
---

Migrated from https://github.com/WICG/spatial-navigation/issues/128
Originally created by @hugoholgersson on *Thu, 01 Nov 2018 20:33:20 GMT*

---
I created a Chromium-bug that describes this in problem in more detail: https://crbug.com/900499.

The current algorithm (pictured in 7.2) does not change the starting point* when climbing to an ancestor container.

\* In Chrome we use the term "search origin" instead of "starting point".

**Comment by frivoal** [comment]
---

Migrated from https://github.com/WICG/spatial-navigation/issues/128#issuecomment-436164659
Originally created by @hugoholgersson on *Tue, 06 Nov 2018 08:07:23 GMT*

---
> if the starting point which is a decendent of the scrollable element becomes out of the view, the starting point needs to be changed to the bounding border box of the scrollable element.

Yes, that's what I mean, but this is _only when searching outside_ the scroll container.

**Comment by frivoal** [comment]
---

Migrated from https://github.com/WICG/spatial-navigation/issues/128#issuecomment-436926230
Originally created by @jihyerish on *Thu, 08 Nov 2018 09:18:35 GMT*

---
Oh, I see. Then I can see two topics here:

[[Topic 1]]()
If the starting point inside the scroll container becomes out of the view, reset the starting point to the scroll container.
[[Topic 2]]()
Use the bounding border box(visual bounding box) of the scroll container only when searching *outside*.

Topic 1 is also related to https://github.com/WICG/spatial-navigation/issues/129.
I think we need to discuss more about it.

But for Topic 2, why do you think that the scope for searching the candidate is the visual bounding box only when the focus moving outside?
Does it mean that if the focus is about to move inside the scroll container and the scroll bar is at the end, the focus will move to the first element in the scroll container which is offscreen?

**Comment by frivoal** [comment]
---

Migrated from https://github.com/WICG/spatial-navigation/issues/128#issuecomment-436942513
Originally created by @hugoholgersson on *Thu, 08 Nov 2018 10:11:47 GMT*

---
>But for Topic 2, why do you think that the scope for searching the candidate is the visual bounding box only when the focus moving outside?

Look at @junhoseo's [test case](https://chromium-review.googlesource.com/c/chromium/src/+/1253523/8/third_party/WebKit/LayoutTests/fast/spatial-navigation/snav-search-outside-of-focused-scroller.html). It demonstrates why we cannot use the scroll container's edge when searching outside of the [focused] scroll container.

> Does it mean that if the focus is about to move inside the scroll container and the scroll bar is at the end, the focus will move to the first element in the scroll container which is offscreen?

Spatnav should never move focus to an offscreen element... When moving towards a scroll container, no matter if it contains visible focusables or not, focus should [first go to the scroll container](http://bit.ly/snav2) itself. (Remember that [scroll containers are focusables too](https://groups.google.com/a/chromium.org/d/msg/blink-dev/Fku6784p0wI/3MyOMTk7BwAJ).)

**Comment by frivoal** [comment]
---

Migrated from https://github.com/WICG/spatial-navigation/issues/128#issuecomment-436155195
Originally created by @jihyerish on *Tue, 06 Nov 2018 07:23:42 GMT*

---
In current spec, the focus moves inside or outside of the scrollable element depending on the bounding border box of the scrollable element.
(The definition of the boundary box in the spec: https://wicg.github.io/spatial-navigation/#boundary-box)

I'm not sure I understand clearly about your point. 
Am I understanding right about your opinion as below?:

- if the starting point which is a decendent of the scrollable element becomes out of the view, the starting point needs to be changed to the bounding border box of the scrollable element.

**Comment by frivoal** [comment]
For anyone trying to read the above, it seems that the issue migration script put the comments out of order. I suggest reading the original https://github.com/WICG/spatial-navigation/issues/128, it's easier to follow.

**Comment by jihyerish** [comment]
@hugoholgersson Sorry for the broken order of the comments. 
For your [last comment](https://github.com/w3c/csswg-drafts/issues/3392#issuecomment-443620629), now I understand your example.
Also, there had been the update in the spec, the term _**starting point**_ was changed to **_search origin_** as same as in the blink.

I think there are two approaches:
* Scenario: 
   - The currently focused element **_A_** (initial **search origin**) inside the scrollable element **_S_** becomes out of the view by scrolling.

(1) Set the **search origin** to the **_S_** as soon as the **_A_** becomes offscreen.
(2) Set the **search origin** to the **_S_** only if the navigation direction indicates searching the candidate outside of the **_S_**.

If the issue #3391 is resolved, it will make easy to decide this issue, too.

I prefer (1), because it is more simple and constant behavior both for searching inside and outside the scrollable element.
@frivoal, what do you think?

**Comment by jihyerish** [comment]
I think the search origin need to change as below when the focus moves outside of the spatial navigation container.
(NOTE: the condition of spatial navigation container = 
     the viewport of a browsing context (not limited to the top-level browsing context) ||
     scroll containers || element specified with `spatial-navigation-contain: contain`)
    
i) Previous search origin is a visible focusable element inside iframe
=> Search origin will change to the document of the iframe

ii) Previous search origin is a visible focusable element inside the spatial navigation container except for iframe
 => Search origin won't change
 => If you look at the figure below:
       
![image](https://user-images.githubusercontent.com/6636090/64580109-1d48c280-d3c0-11e9-988c-b5443c2d34a8.png)
The element with blue outline is the currently focused element and the search origin.
If you press the down arrow key to move the focus out of the container, it looks natural when the focus moves to 'B' . And this is the case that the search origin doesn't change.
However, if the search origin changes to the container before the focus moves out of the container, the focus will go to 'A' when you press the down arrow key, and I think it isn't reasonable behavior.

Also, I think the search origin should change only when the directional input used for spatial navigation, not scrolling.
(Current browser behavior: when there is a focused element **_A_** inside the scroller **_S_** and after scrolling **_S_** with arrow keys, the **_A_** is still the focused element.)

@hugoholgersson, what do you think?

**Comment by hugoholgersson** [comment]
> If you press the down arrow key to move the focus out of the container, it looks natural when the focus moves to 'B' . And this is the case that the search origin doesn't change.

Makes sense, yes! 

> @hugoholgersson, what do you think?

- When activeElement is **not completely scrolled out** (it's still [partially] visible in its scrollport S):
  - Always search from activeElement's visible part.

- When activeElement is **completely scrolled out** from its scrollport S:
  1. First search in S from S's [edge](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/page/spatial_navigation.cc?l=678&gsn=OppositeEdge).
  2. If nothing was found or S couldn't scroll further: search from S's box.

(This is [what I first suggested](https://crbug.com/900499) but I've added the condition for when activeElement is still visible, like the node with blue outline in your picture above.)

When activeElement is not scrolled out (it's still [partially] visible in its scrollport S) then, of course, we should keep the visible part of activeElement as search origin. I've added one if-statement to https://bit.ly/snav2 so you see this rule in the bigger picture.
